segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/exts3/ExtS3Storage.java

```diff
@@ -0,0 +1,445 @@
+        return null;
+    }
+
+    private Void syncConcat(SegmentHandle targetHandle, long offset, String sourceSegment) throws StreamSegmentNotExistsException {
+        long traceId = LoggerHelpers.traceEnter(log, "concat", targetHandle.getSegmentName(),
+                offset, sourceSegment);
+
+        SortedSet<MultipartPartETag> partEtags = new TreeSet<>();
+        String targetPath = config.getExts3Root() + targetHandle.getSegmentName();
+        String uploadId = client.initiateMultipartUpload(config.getExts3Bucket(), targetPath);
+
+        // check whether the target exists
+        if (!syncExists(targetHandle.getSegmentName())) {
+            throw new StreamSegmentNotExistsException(targetHandle.getSegmentName());
+        }
+        // check whether the source is sealed
+        SegmentProperties si = syncGetStreamSegmentInfo(sourceSegment);
+        if (!si.isSealed()) {
+            throw new IllegalStateException(sourceSegment);
+        }
+
+        //Upload the first part
+        CopyPartRequest copyRequest = new CopyPartRequest(config.getExts3Bucket(),
+                targetPath,
+                config.getExts3Bucket(),
+                targetPath,
+                uploadId,
+                1).withSourceRange(Range.fromOffsetLength(0, offset));
+        CopyPartResult copyResult = client.copyPart(copyRequest);
+
+        partEtags.add(new MultipartPartETag(copyResult.getPartNumber(), copyResult.getETag()));
+
+        //Upload the second part
+        S3ObjectMetadata metadataResult = client.getObjectMetadata(config.getExts3Bucket(),
+                config.getExts3Root() + sourceSegment);
+        long objectSize = metadataResult.getContentLength(); // in bytes
+
+        copyRequest = new CopyPartRequest(config.getExts3Bucket(),
+                config.getExts3Root() + sourceSegment,
+                config.getExts3Bucket(),
+                targetPath,
+                uploadId,
+                2).withSourceRange(Range.fromOffsetLength(0, objectSize));
+
+        copyResult = client.copyPart(copyRequest);
+        partEtags.add(new MultipartPartETag(copyResult.getPartNumber(), copyResult.getETag()));
+
+        //Close the upload
+        client.completeMultipartUpload(new CompleteMultipartUploadRequest(config.getExts3Bucket(),
+                targetPath, uploadId).withParts(partEtags));
+
+        client.deleteObject(config.getExts3Bucket(), config.getExts3Root() + sourceSegment);
+        LoggerHelpers.traceLeave(log, "concat", traceId);
+
+        return null;
+    }
+
+    private Void syncDelete(SegmentHandle handle) {
+
+        client.deleteObject(config.getExts3Bucket(), config.getExts3Root() + handle.getSegmentName());
+        return null;
+    }
+
+    /**
+     * Executes the given supplier asynchronously and returns a Future that will be completed with the result.
+     */
+    private <R> CompletableFuture<R> supplyAsync(String segmentName, Callable<R> operation) {
```

Comment:
This is the same as in `FileSystemStorage`. Please extract the method somewhere and reuse it.